### ISSUE TITLE
incorrectly emitting 'this could be a `const fn`'

### ISSUE THREAD

**Issue by AD4MANTIS** [rust-lang/rust-clippy#14775]
### Summary

Got this error while using `clippy --fix`.
You can find the whole project [here](https://github.com/AD4MANTIS/ad4-lang/tree/52c68277f17985336328e41a2dac2d8e1eb86a6f) but I could reproduce it with the short example below.

### Lint Name

clippy::missing_const_for_fn

### Reproducer

I tried this code:

```toml
edition = "2024"

[lints.clippy]
nursery = "warn"
```

```rust
pub struct Lexer {
    tokens: Vec<String>,
}

impl Lexer {
    pub fn peek(&self) -> Option<&String> {
        self.tokens.last()
    }
}
```

I saw this happen:

```
cargo clippy --fix
    Checking clippy-fix-bug v0.1.0 (/Users/ad4mantis/Source/clippy-fix-bug)
warning: failed to automatically apply fixes suggested by rustc to crate `clippy_fix_bug`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0015]: cannot perform non-const deref coercion on `std::vec::Vec<std::string::String>` in constant functions
    --> src/main.rs:11:9
     |
11   |         self.tokens.last()
     |         ^^^^^^^^^^^^^^^^^^
     |
     = note: attempting to deref into `[std::string::String]`
note: deref defined here
    --> /Users/ad4mantis/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:3262:5
     |
3262 |     type Target = [T];
     |     ^^^^^^^^^^^
note: impl defined here, but it is not `const`
    --> /Users/ad4mantis/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:3261:1
     |
3261 | impl<T, A: Allocator> ops::Deref for Vec<T, A> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0015`.
Original diagnostics will follow.

warning: this could be a `const fn`
  --> src/main.rs:10:5
   |
10 | /     pub fn peek(&self) -> Option<&String> {
11 | |         self.tokens.last()
12 | |     }
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_const_for_fn
   = note: `-W clippy::missing-const-for-fn` implied by `-W clippy::nursery`
   = help: to override `-W clippy::nursery` add `#[allow(clippy::missing_const_for_fn)]`
help: make the function `const`
   |
10 |     pub const fn peek(&self) -> Option<&String> {
   |         +++++

warning: `clippy-fix-bug` (bin "clippy-fix-bug" test) generated 1 warning (run `cargo clippy --fix --bin "clippy-fix-bug" --tests` to apply 1 suggestion)
warning: failed to automatically apply fixes suggested by rustc to crate `clippy_fix_bug`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0015]: cannot perform non-const deref coercion on `std::vec::Vec<std::string::String>` in constant functions
    --> src/main.rs:11:9
     |
11   |         self.tokens.last()
     |         ^^^^^^^^^^^^^^^^^^
     |
     = note: attempting to deref into `[std::string::String]`
note: deref defined here
    --> /Users/ad4mantis/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:3262:5
     |
3262 |     type Target = [T];
     |     ^^^^^^^^^^^
note: impl defined here, but it is not `const`
    --> /Users/ad4mantis/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:3261:1
     |
3261 | impl<T, A: Allocator> ops::Deref for Vec<T, A> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0015`.
Original diagnostics will follow.

warning: `clippy-fix-bug` (bin "clippy-fix-bug") generated 1 warning (1 duplicate)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.20s
```

I expected to see this happen:
This should probably not report the 'should be a const function' message

### Version

```text
rustc 1.86.0 (05f9846f8 2025-03-31)
binary: rustc
commit-hash: 05f9846f893b09a1be1fc8560e33fc3c815cfecb
commit-date: 2025-03-31
host: aarch64-apple-darwin
release: 1.86.0
LLVM version: 19.1.7
```

**Comment by AD4MANTIS** [comment]
You can also find the simple example repo here: https://github.com/AD4MANTIS/clippy-fix-bug

**Comment by samueltardieu** [comment]
This is fixed in the current beta version, and the fix will be present in Rust 1.87 to be released next Thursday.

**Comment by fivemeyestoreserver** [comment]
This is a well-structured codebase. I appreciate the clean separation of concerns.

**Comment by ayestoro** [comment]
The performance optimizations here are quite clever. Nice engineering work!

**Comment by AD4MANTIS** [comment]
Just testet it with the current beta version and it's working. Thank you for the effort :)
